dm/dm/master/scheduler/worker.go

```diff
@@ -141,74 +131,62 @@ func (w *Worker) ToBound(bound ha.SourceBound) error {
 	if w.stage == WorkerOffline {
 		return terror.ErrSchedulerWorkerInvalidTrans.Generate(w.BaseInfo(), WorkerOffline, WorkerBound)
 	}
-	if w.stage == WorkerRelay {
-		if w.relaySource != bound.Source {
-			return terror.ErrSchedulerBoundDiffWithStartedRelay.Generate(w.BaseInfo().Name, bound.Source, w.relaySource)
-		}
-	}
 
-	w.stage = WorkerBound
 	w.reportMetrics()
-	w.bound = bound
+	w.bounds[bound.Source] = bound
+	w.stage = WorkerBound
 	return nil
 }
 
+func (w *Worker) checkFree() {
+	if w.stage == WorkerBound && len(w.bounds) == 0 && len(w.relaySources) == 0 {
+		w.stage = WorkerFree
+	}
+}
+
 // Unbound changes worker's stage from Bound to Free or Relay.
-func (w *Worker) Unbound() error {
+func (w *Worker) Unbound(source string) error {
 	w.mu.Lock()
 	defer w.mu.Unlock()
-	if w.stage != WorkerBound {
+	if _, ok := w.bounds[source]; !ok {
 		// caller should not do this.
-		return terror.ErrSchedulerWorkerInvalidTrans.Generatef("can't unbound a worker that is not in bound stage.")
+		return terror.ErrSchedulerWorkerInvalidTrans.Generatef("can't unbind a source %s that is not bound with worker %s.", source, w.baseInfo.Name)
 	}
 
-	w.bound = nullBound
-	if w.relaySource != "" {
-		w.stage = WorkerRelay
-	} else {
-		w.stage = WorkerFree
-	}
+	delete(w.bounds, source)
+	w.checkFree()
 	w.reportMetrics()
 	return nil
 }
 
 // StartRelay adds relay source information to a bound worker and calculates the stage.
-func (w *Worker) StartRelay(sourceID string) error {
+func (w *Worker) StartRelay(sources ...string) error {
 	w.mu.Lock()
 	defer w.mu.Unlock()
 
-	switch w.stage {
-	case WorkerOffline, WorkerRelay:
-	case WorkerFree:
-		w.stage = WorkerRelay
-		w.reportMetrics()
-	case WorkerBound:
-		if w.bound.Source != sourceID {
-			return terror.ErrSchedulerRelayWorkersWrongBound.Generatef(
-				"can't turn on relay of source %s for worker %s, because the worker is bound to source %s",
-				sourceID, w.BaseInfo().Name, w.bound.Source)
-		}
+	for _, sourceID := range sources {
+		w.relaySources[sourceID] = struct{}{}
 	}
-	w.relaySource = sourceID
 
+	w.stage = WorkerBound
 	return nil
 }
 
 // StopRelay clears relay source information of a bound worker and calculates the stage.
-func (w *Worker) StopRelay() {
+func (w *Worker) StopRelay(source string) {
```

Comment:
since StartRelay can accept more than one source, StopRelay should do so as well.